riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+    link => byte2.check(tmp4_h);
+    link => byte2.check(tmp4_l);
+    link => byte2.check(tmp5_h);
+    link => byte2.check(tmp5_l);
+    link => byte2.check(tmp6_h);
+    link => byte2.check(tmp6_l);
+
+    // We need to add these inline instead of using std::utils::is_zero
+    // because when XX is not constrained, witgen will try to set XX,
+    // XX_inv and XXIsZero to zero, which fails this constraint.
+    // Therefore, we have to activate the following constraint whenever XXIsZero is used:
+    // XXIsZero = 1 - XX * XX_inv
+    col witness XX, XX_inv, XXIsZero;
+    std::utils::force_bool(XXIsZero);
+    XXIsZero * XX = 0;
+
+    // ============== control-flow instructions ==============
+
+    // Load the value of label `l` into register X.
+    // We restrict the address to 24 bits to avoid overflows.
+    instr load_label XL, l: label
+        link ~> regs.mstore(0, XL, STEP, tmp1_h, tmp1_l)
+        link => byte.check(tmp1_h)
+    {
+        tmp1_h * 2**16 + tmp1_l = l
+    }
+
+    // Jump to `l` and store the return program counter in register W.
+    // We restrict the address to 24 bits to avoid overflows.
+    instr jump l: label, WL
+        link ~> regs.mstore(0, WL, STEP, tmp1_h, tmp1_l)
+        link => byte.check(tmp1_h)
+    {
+        pc + 1 = (tmp1_h * 2**16) + tmp1_l,
+        pc' = l
+    }
+    
+    // Jump to the address in register X and store the return program counter in register W.
+    instr jump_dyn XL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> regs.mstore(0, WL, STEP, tmp2_h, tmp2_l)
+        // pc is capped at 24 bits, so for this instruction 
+        // we restrict the higher limbs to 1 byte
+        link => byte.check(tmp1_h)
+        link => byte.check(tmp2_h)
+    {
+        // We need a temporary col here because the compiler doesn't like
+        // `exp` inside an expr on the RHS of a pc update. 
+        pc' = (tmp1_h * 2**16) + tmp1_l,
+        pc + 1 = tmp2_h * 2**16 + tmp2_l
+    }
+
+    // Jump to `l` if val(X) != val(Y) is nonzero, where X and Y are register ids.
+    instr branch_if_not_equal XL, YL, l: label
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link => byte.check(tmp3_h)
+        link => byte.check(tmp3_l)
+        link => byte.check(tmp4_h)
+        link => byte.check(tmp4_l)
+        link => byte.check(tmp5_h)
+        link => byte.check(tmp5_l)
+        link => byte.check(tmp6_h)
+        link => byte.check(tmp6_l)
+    {
+        XXIsZero = 1 - XX * XX_inv,
+
+        // What we want to do is to compare tmp1_h == tmp2_h && tmp1_l == tmp2_l.
+        // This could be done with the constraint below, but overflows causes soundness issues:
+        // XX = (tmp1_h - tmp2_h)**2 + (tmp1_l - tmp2_l)**2,
+        // One solution is to further decompose into single bytes.
+
+        // Decompose tmp1_h into 2 bytes
+        tmp1_h = tmp3_h * 2**8 + tmp3_l,
```

Comment:
This is a good example of why we should allow links inside bodies, so that the range checks are next to these decompositions. Not for this PR, but is there anything blocking this?

Reply:
No this is working, just wanted more people to check soundness. The byte range checks are in the links above also.

Reply:
Oh I don't mean that it doesn't work, just that I had to jump back up to the links the check the range checks and it would be great they were here instead. Unrelated to this PR.

Reply:
ah ok